data_manipulation.py

```diff
@@ -1,42 +1,46 @@
 import numpy as np
 import pandas as pd
-
 from sklearn import preprocessing
 import typing
 
 
-def prepare_training_data(train_data: pd.DataFrame) -> typing.Tuple[np.ndarray, preprocessing.MinMaxScaler]:
+def generate_input_data(train_data: pd.DataFrame, input_str: str) -> typing.Tuple[np.ndarray, preprocessing.MinMaxScaler]:
+
+    input_scaler = preprocessing.MinMaxScaler()
+    input_data = train_data[input_str].values.astype(float)
+    input_data = np.array(input_data).reshape((len(input_data), 1))
+    input_data = input_scaler.fit_transform(input_data)
+    return input_data, input_scaler
+
+
+def prepare_training_data(train_data: pd.DataFrame, extra_input_list_str=[]) -> typing.Tuple[np.ndarray, preprocessing.MinMaxScaler]:
     """
     Receive the data for training the model.
     All the data will be normalized to a value between -1 and 1 to make it easier to train the model.
     The function returns the following values in a tuple:
     :return: data: a numpy array containing 3 columns: sentiment, stock value at closing time, free cash flow
     :return: *_scaler: the MinMaxScaler used to normalize the data. Used to transform the data back to normal
     """
-    sentiments_scaler = preprocessing.MinMaxScaler()
-    close_scaler = preprocessing.MinMaxScaler()
-    close_p_scaler = preprocessing.MinMaxScaler()
-    fcf_scaler = preprocessing.MinMaxScaler()
-
-    sentiments = train_data['Sentiment'].values.astype(float)
-    sentiments = np.array(sentiments).reshape((len(sentiments), 1))
-    sentiments = sentiments_scaler.fit_transform(sentiments)
 
     close = train_data['close'].values.astype(float)
     close = np.array(close).reshape((len(close), 1))
-
-    close_p = (close[1:] - close[:-1]) / close[1:]
-    close_p = close_p_scaler.fit_transform(close_p)
-
+    close_scaler = preprocessing.MinMaxScaler()
     close = close_scaler.fit_transform(close)
+    # close_p_scaler = preprocessing.MinMaxScaler()
+    # close_p = (close[1:] - close[:-1]) / close[1:]
+    # close_p = close_p_scaler.fit_transform(close_p)
+
+    extra_input_list = []
+    extra_scaler_list = []
+    if len(extra_input_list_str) != 0:
+        for input in extra_input_list_str:
+            extra_input, extra_scaler = generate_input_data(train_data, input)
+            extra_input_list = extra_input_list + [extra_input[1:]]
+            extra_input_scaler = extra_scaler_list + [extra_scaler]
 
-    fcf = train_data['fcf'].values.astype(float)
-    fcf = np.array(fcf).reshape((len(fcf), 1))
-    fcf = fcf_scaler.fit_transform(fcf)
+    data_to_concat = tuple(extra_input_list)
```

Comment:
I don't think you have to convert your list into a tuple

Reply:
It should be ok after discussion